File: finance-ui/src/lib/api/coa.ts
import { apiFetch } from './client';

export type CoaAccountItem = {
  id: string;
  account_code: string;
  account_name: string;
  account_subtype?: string | null;
};

export function listCoaAccounts(params?: { account_type?: string; is_active?: boolean }) {
  const q = new URLSearchParams();
  if (params?.account_type) q.set('account_type', params.account_type);
  if (params?.is_active != null) q.set('is_active', String(params.is_active));
  const suffix = q.toString() ? `?${q}` : '';
  return apiFetch<CoaAccountItem[]>(`/coa-accounts${suffix}`);
}

File: accfin/app/api/routes/coa.py
"""Chart of Accounts API — expense account lookup for parsing confirmation UI."""

from uuid import UUID

from fastapi import APIRouter, Depends, Query
from pydantic import BaseModel, ConfigDict
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.database import get_db_session
from app.core.dependencies import require_permission
from app.models.ledger import CoaAccount
from app.schemas.auth import TokenData

router = APIRouter(prefix="/coa-accounts", tags=["Chart of Accounts"])


class CoaAccountResponse(BaseModel):
    id: UUID
    account_code: str
    account_name: str
    account_subtype: str | None = None

    model_config = ConfigDict(from_attributes=True)


@router.get("", response_model=list[CoaAccountResponse])
async def list_coa_accounts(
    account_type: str | None = Query(None),
    is_active: bool = Query(True),
    session: AsyncSession = Depends(get_db_session),
    _: TokenData = Depends(require_permission("expenses:read")),
) -> list[CoaAccountResponse]:
    q = select(CoaAccount).where(CoaAccount.is_active == is_active)
    if account_type:
        q = q.where(CoaAccount.account_type == account_type)
    q = q.order_by(CoaAccount.account_code)
    result = await session.execute(q)
    return list(result.scalars().all())
